pkg/registry/core/service/ipallocator/cidrallocator.go

```diff
@@ -0,0 +1,413 @@
+		return err
+	}
+
+	cidrsSet := sets.New[string]()
+	cidrState := map[string]bool{}
+	for _, cidr := range serviceCIDRs {
+		ready := true
+		if !isReady(cidr) || !cidr.DeletionTimestamp.IsZero() {
+			ready = false
+		}
+
+		if c.ipFamily == api.IPv4Protocol &&
+			cidr.Spec.IPv4 != "" {
+			cidrsSet.Insert(cidr.Spec.IPv4)
+			cidrState[cidr.Spec.IPv4] = ready
+		}
+		if c.ipFamily == api.IPv6Protocol &&
+			cidr.Spec.IPv6 != "" {
+			cidrsSet.Insert(cidr.Spec.IPv6)
+			cidrState[cidr.Spec.IPv6] = ready
+		}
+	}
+
+	// obtain the existing allocators and set the existing state
+	treeSet := sets.New[string]()
+	c.muTree.Lock()
+	c.tree.Walk(func(k netip.Prefix, v *Allocator) bool {
+		v.ready.Store(cidrState[k.String()])
+		treeSet.Insert(k.String())
+		return false
+	})
+	c.muTree.Unlock()
+	cidrsToRemove := treeSet.Difference(cidrsSet)
+	cidrsToAdd := cidrsSet.Difference(treeSet)
+
+	// Remove allocators that no longer exist
+	errs := []error{}
+	for _, cidr := range cidrsToRemove.UnsortedList() {
+		c.deleteAllocator(cidr)
+	}
+
+	// Add new allocators
+	for _, cidr := range cidrsToAdd.UnsortedList() {
+		_, ipnet, err := netutils.ParseCIDRSloppy(cidr)
+		if err != nil {
+			return err
+		}
+		// New ServiceCIDR, create new allocator
+		allocator, err := NewIPAllocator(ipnet, c.client, c.ipAddressInformer)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		allocator.ready.Store(cidrState[cidr])
+		c.addAllocator(cidr, allocator)
+		klog.Infof("Created ClusterIP allocator for Service CIDR %s", cidr)
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+func (c *CIDRAllocator) getAllocator(ip net.IP) (*Allocator, error) {
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+	mask := 32
+	if c.ipFamily == api.IPv6Protocol {
+		mask = 128
+	}
+	key := fmt.Sprintf("%s/%d", ip.String(), mask)
+	// Use the largest subnet to allocate addresses
+	_, allocator, ok := c.tree.ShortestPrefixMatch(key)
+	if !ok {
+		return nil, ErrMismatchedNetwork
+	}
+	return allocator, nil
+}
+
+func (c *CIDRAllocator) addAllocator(cidr string, allocator *Allocator) {
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+	c.tree.Insert(cidr, allocator)
+}
+
+func (c *CIDRAllocator) deleteAllocator(cidr string) {
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+	_, ok := c.tree.Delete(cidr)
+	if ok {
+		klog.V(3).Infof("CIDR %s deleted", cidr)
+	}
+}
+
+func (c *CIDRAllocator) AllocateService(service *api.Service, ip net.IP) error {
+	allocator, err := c.getAllocator(ip)
+	if err != nil {
+		return err
+	}
+	return allocator.AllocateService(service, ip)
+}
+
+func (c *CIDRAllocator) Allocate(ip net.IP) error {
+	allocator, err := c.getAllocator(ip)
+	if err != nil {
+		return err
+	}
+	return allocator.Allocate(ip)
+}
+
+func (c *CIDRAllocator) AllocateNextService(service *api.Service) (net.IP, error) {
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+
+	// TODO add strategy to return the allocator with more free addresses
+	for _, allocator := range c.tree.Parents() {
```

Comment:
the allocators randomize the IPs assigned, I don't think that deterministic is a good strategy, what I'm trying to mean is that it should be weighted, i.e. if I have 3 allocators with 256, 1024 and 4096 addresses each, we should make first allocator less likely to hit

Reply:
however, I can see how this can change in the future, so we may add an Strategy like we added in the bitmap allocators, I will add a TODO to revisit on the next iteration

Reply:
This will need to be revisited before beta